stl/inc/xutility

```diff
@@ -4767,7 +4757,7 @@ _NODISCARD bool equal(const _InIt1 _First1, const _InIt1 _Last1, const _InIt2 _F
     const auto _UFirst2 = _Get_unwrapped_n(_First2, _Idl_distance<_InIt1>(_UFirst1, _ULast1));
     return _Equal_unchecked(_UFirst1, _ULast1, _UFirst2, _Pass_fn(_Pred));
 }
-#endif // _HAS_IF_CONSTEXPR
+#endif // !_HAS_IF_CONSTEXPR

```

Comment:
Is this what we typically do?  If you look at the example just below (line 4793-4871 before changes), the `#endif` is commented with the original condition (not negated).  Doing a quick search throughout the code base also shows that the original comment follows our convention.

Reply:
As far as I know there was a discussion early after open sourcing and I have the memory that the majority was in favor of 

```cpp
#if foo
#else // ^^^ foo / !foo
#endif // !foo
```

The reason was that for longer clauses such as this just reading `#endif // foo` would lead you to believe that foo is active or require you to check the start of the block.

That said sleep was rather short this year so I doubt any memory I have

Reply:
This is #351 - existing code has accumulated inconsistent patterns. I'm not sure how we ended up with the "hybrid arrow" examples that @mnatsuhara pointed out - I think they evolved from our classic pattern of

```cpp
#if _HAS_CATS
#else // _HAS_CATS
#endif // _HAS_CATS

#ifdef __cpp_lib_kittens
#else // __cpp_lib_kittens
#endif // __cpp_lib_kittens
```

by replacing just the middle portion with arrows.

My personal opinion: Aside from the question of when comments can be omitted, I'd prefer for us to standardize on consistent arrow comments:

```cpp
#if _HAS_CATS
#else // ^^^ _HAS_CATS ^^^ / vvv !_HAS_CATS vvv
#endif // ^^^ !_HAS_CATS ^^^

#ifdef __cpp_lib_kittens
#else // ^^^ defined(__cpp_lib_kittens) ^^^ / vvv !defined(__cpp_lib_kittens) vvv
#endif // ^^^ !defined(__cpp_lib_kittens) ^^^
```

(I'm not very concerned with whether ` ^^^ / vvv ` or just ` / ` appears in the middle; I like the former slightly more.) I prefer this because it's the clearest in dense regions of preprocessor logic, and it's also unambiguous about whether the final comment should say `_HAS_CATS` or `!_HAS_CATS` (because it points to a specific section of code).

I could see an argument that "of course an `#endif` closes a section of code, so we just need to consistently identify what we closed", which would argue for `!_HAS_CATS` here.

Somewhat related: I believe we've accumulated some inconsistency about how to refer to `#ifdef` (and `#if defined(MEOW)` etc.) conditions in comments - I have seen some comments switch from the `defined(MEOW)` / `!defined(MEOW)` domain to the `MEOW` / `!MEOW` domain which I find confusing. This is because of the danger of actually confusing the definedness versus valueness domains in the test (and the compiler warning about `#if NOT_DEFINED` being treated as `#if 0`).

There are "only" 444 occurrences of `#else` in `stl/inc`, but 2049 `#endif`s - tedious to audit, but still easier than bracing all control flow, which we finally did and it was awesome. I should probably just submit a PR to prevent this from burning up valuable time in the future. :fire:

Reply:
I have to agree that we should be consistent. As I had to rebase anyway I removed that change so that it can be tackled with all the other comments in a consistent fashion

Reply:
Maybe this would be a "good second issue"